rosidl_runtime_py/rosidl_runtime_py/set_message.py

```diff
@@ -0,0 +1,57 @@
+# Copyright 2017-2019 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Any
+from typing import Dict
+
+
+class SetFieldError(Exception):
+
+    def __init__(self, field_name, exception):
+        super(SetFieldError, self).__init__()
+        self.field_name = field_name
+        self.exception = exception
+
+
+def set_message_fields(msg: Any, values: Dict[str, str]) -> None:
+    """
+    Set the fields of a ROS message.
+
+    :param msg: The ROS message to populate.
+    :param values: The values to set in the ROS message. The keys of the dictionary represent
+        fields of the message.
+    :raises SetFieldError: If there is an error setting a message field (e.g. the message does
+        not have a field provided in the input dictionary).
+    """
+    for field_name, field_value in values.items():
+        try:
+            field_type = type(getattr(msg, field_name))
+        except Exception as e:
+            raise SetFieldError(field_name, e)
+        try:
+            value = field_type(field_value)
+        except TypeError:
+            value = field_type()
+            try:
+                set_message_fields(value, field_value)
```

Comment:
Calling this recursively in case of a `TypeError` seems weird to me. This should only happen when the field type is actually another message.

Reply:
From the original implementation. How can we check if the subtype is a message?

Reply:
Each message class has a class method [get_fields_and_field_types](https://github.com/ros2/rosidl_python/blob/4f11d8734a5b78daf874a680359e813d42ca5fde/rosidl_generator_py/resource/_msg.py.em#L312) which provides the type of each slot.

But since it has worked this doesn't need to be changed in this PR (also that API is likely going to change soon, see #33).